### ISSUE TITLE
Vexflow 5 release plan

### ISSUE THREAD

**Issue by ronyeh** [vexflow/vexflow#224]
Let's make a concrete plan to get 5.0 released. 

The previous discussion was at #13.

**Comment by ronyeh** [comment]
I propose for version 5.0:
- prioritize fixes and features needed by @rvilarl and @jaredjj3 
- make sure the readme and hello world tutorials make sense
- have a simple sandbox hosted on codepen/jsfiddle/etc so people can quickly test out vexflow snippets. 

5.0 will have API differences from 4.x. That's ok. We can improve backwards compatibility as we move forward to 5.1, 5.2, ...as the need arises. 

For 5.1:
- prioritize fixes and features needed by @AaronDavidNewman 

For 5.2:
- prioritize fixes and features needed by @mscuthbert 

For 5.3:
- prioritize fixes and features needed by the OSMD project. They are still using (a modified) version of VexFlow 1.x because they prefer its formatter. We can perhaps figure out the formatting puzzle once and for all 😅. (The solution might just be to allow for switching between multiple formatting algorithms.)

I propose this approach because these are the users most interested in and invested in VexFlow. Let's make our users happy.

**Comment by ronyeh** [comment]
@rvilarl regarding API docs:

Yes we should generate them. They should be stored in versioned directories in the **vexflow-docs** repository. 

So the directory structure could be:

```
vexflow-docs/
  ver/
    5.0.0/
    5.1.0/
    ...
  2024/
    0724-f8b49c2/
    0816-97da26e/
    ...
```

This allows us to manually generate the API docs for releases, or for whenever we want to keep an intermediate version around (like if we issued a beta on July 24 with short hash f8b49c2). 

Having versioned API docs is useful. Our solution before was nice but it only ever matched the head revision.

**Comment by jsmcx** [comment]
Thank you for highlighting the OSMD project. As a consumer of both libraries, it always seemed to me that it would be best to have commonalities in the formatting code, so that improvements could benefit everyone.

**Comment by rvilarl** [comment]
For 5.0, I would like to close
https://github.com/0xfe/vexflow/issues/573
#220 
#223 
@jaredjj3 you?

**Comment by mscuthbert** [comment]
#220 seems well on its way, so I understand wanting to get it into v.5

#223 is a newer issue that has come up for the first time later in the v5 development.  Could it go into a 5.1 branch issue?

https://github.com/0xfe/vexflow/issues/573 is an issue that has been open since 2017 with no movement or comment for 7 years on it -- why is it suddenly a priority to get in before the next release?  

I understand that there are things each of us want in a release, but at present Vexflow is in a stasis position -- people who aren't interested in the font development (like me and several others who have expressed concern) can't contribute to the project until a stable release with the font changes has come out.  In the meantime some people for whom fonts are most important are proposing more changes (or APIs to be documented etc.) that are necessary before another release can come about.

In the meantime, contributions to other parts of the library (formatting, layouts, beam improvements) are languishing.  My feeling is that a feature freeze for V5 is long overdue (I expressed this feeling to Ron in December of last year) and anything else people want should be in 5.1, 5.2, or (if significantly backwards incompatible) 6.0.

**Comment by mscuthbert** [comment]
> I propose for version 5.0:
> 
> * prioritize fixes and features needed by @rvilarl and @jaredjj3
> * make sure the readme and hello world tutorials make sense
> * have a simple sandbox hosted on codepen/jsfiddle/etc so people can quickly test out vexflow snippets.

I think that the second of these is the only thing to make sure to have for v5.  Making sure that the new font-system can be used in the variety of use cases that Vexflow has been in for years (including posting images to GitHub for review without assuming users have certain fonts installed) is the main other thing I see missing for v5.  

I'm happy to bump the 5.1, 5.2, 5.3 back one 0.1 release to make the other two issues (prioritize fixes and features needed by @rvilarl and @jaredjj3  -- and the sandbox) be the new 5.1.  

----
Though I think that I haven't been clear about what was useful about the old sandbox -- it let a developer see how Vexflow was running on her or his own development build (that hadn't yet been committed or even PR'd).  I don't see how a JSFiddle version of a sandbox substitutes for this feature since it will never take into account the most recent, say beam angle tweak, that a developer had just implemented.

**Comment by rvilarl** [comment]
@mscuthbert this is the reason for the 3 PRs that I selected
> #220 seems well on its way, so I understand wanting to get it into v.5

#220 was your proposal, and I find it good. I am waiting on a review to confirm that I understood it properly
 
> #223 is a newer issue that has come up for the first time later in the v5 development. Could it go into a 5.1 branch issue?

#223 is a follow up from #220 as doing the change I realised that the styles were applied more than once on some elements and none on others.

> [0xfe/vexflow#573](https://github.com/0xfe/vexflow/issues/573) is an issue that has been open since 2017 with no movement or comment for 7 years on it -- why is it suddenly a priority to get in before the next release?

This one is a result of reading style topics in VexFlow4 to check if I was missing something related to the topic. We have now the `MetricsDefaults` and `Metrics.getStyle()` I was wondering if it should be `ThemeDefaults` and `Theme.getDefaultStyle()` to align with this PR and perhaps make a `Theme.setDefaultStyle()`

**None of the above is a must for 5.0 I am also looking forward to releasing.**

**Comment by ronyeh** [comment]
For now I'd refer to latest alpha at https://cdn.jsdelivr.net/npm/vexflow@alpha/build/cjs/vexflow.js

Hopefully it's not too much hassle. I'd like the README & examples to "always be correct" as much as possible. If we used a release url with 5.0.0, it won't work and someone dropping by will just think our library is broken.

Once we're ready to promote to beta and full release, we can go back and update those README & examples.

**Comment by rvilarl** [comment]
@mscuthbert @ronyeh @AaronDavidNewman @jaredjj3 I would suggest to make a fix date for the release. If we do not get into 5.0 we will never start getting feedback from other users.

**Comment by rvilarl** [comment]
@ronyeh #225 #226 and #227 are not required to proceed. They solve style related issues. I have label with 5.0 the only two relevant PRs.

Quite a while ago I suggested using opentypejs https://github.com/0xfe/vexflow/issues/1399 but it was prefered not to go this way. This is why we moved to use the fonts directly.

I suggest that we list the different use cases and consider pros and cons and even different approaches. I have in mind two use cases but I am sure that this is not the full picture:
- usage of vexflow in a web page, like my piano app.
- producing SVGs to redistribute them.

**Comment by jsmcx** [comment]
As I've mentioned before, my main use is generating SVG for embedding in websites, Word documents, and PDFs. I'm obliged to "hot desk", so I can't guarantee availability of a music font on my machine, nor those I'm distributing to. So embedding the glyphs in some way is really important. 

Options that have been proposed so far:

1. The Vexflow 4 approach of having each glyph as an SVG path element
    Pros: has worked reliably for many years
    Cons: file sizes are bigger than other approaches

2. Use real music fonts
    Pros: Smaller files, easier to change to different font
   Con: Display breaks if person viewing the file doesn't have the font installed. 

3. Embed the font (or a subset) in the generated SVG
    Pros: Files remain relatively small (beyond initial cost of embedded glyphs) 
    Cons: Doesn't render reliably on sites with a strict Content Security Policy. Ironically, GitHub has such a security policy, so demo SVGs may not render if hosted here. Also, use of the `<glyph>` tag is [deprecated](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/glyph), so current proposed method would need to be reworked to avoid that

4. Embed glyph paths as symbols, and then instantiate those paths as needed with '<use>'
    Pro: Files remain relatively small. Doesn't rely on styles or fonts, so should render correctly everywhere.
    Cons: May be more technically complex to change fonts

Incidentally, I suspect the other main use case is rendering within apps, as many of the contributors [to this old discussion](https://github.com/0xfe/vexflow/issues/520) seemed to be doing. In such cases a smaller rendered SVG will help performance, so options 2-4 are all better than the Vexflow 4 approach, but there may be other pros and cons in that scenario that I haven't thought of.

**Comment by rvilarl** [comment]
@jsmcx there is another option for SVGs:
5. Embed the font URL in the SVG, this is good to embed in websites but probably not for documents and PDFs.

I believe that a postprocessoing option for SVGs might be acceptable. With a python script as I proposed in https://github.com/vexflow/vexflow/issues/71#issuecomment-2257693914 or using opentypejs.

**Comment by rvilarl** [comment]
@ronyeh from my point of view, ready for a beta

**Comment by jsmcx** [comment]
I would view the need for postprocessing as a significant regression. Generated SVGs should be viewable stand-alone with no additional work by the user. Complexity should be handled by the library.

**Comment by rvilarl** [comment]
> I would view the need for postprocessing as a significant regression. Generated SVGs should be viewable stand-alone with no additional work by the user. Complexity should be handled by the library.

I believe we should have different options for the different use cases:
- a web page with several SVGs should probably use font-face to download the fonts once in the html header
- SVGs could embed the font URL using font-face
- SVGs could embed the complete font in base64
- SVGs could embed the used subset of the font in base64, see #236
- SVGs could embed the paths using def
- SVGs could use paths rather than text

I believe that the last 5 options could be resolved with scripts. These scripts could be instantiated automatically.

**Comment by ronyeh** [comment]
Let's direct the SVG related discussion over to: https://github.com/vexflow/vexflow/issues/71

**Comment by ronyeh** [comment]
@rvilarl are there any other PRs coming? If not, we can ship a beta this week.

**Comment by rvilarl** [comment]
I will not add any PR after #225. Including #225 is not a must.

**Comment by ronyeh** [comment]
Shipped beta 0.

`npm install vexflow@beta`

Next steps:
- improve examples both in the main repo and over at https://github.com/vexflow/vexflow-examples/
- document major changes and limitations on the Wiki at https://github.com/vexflow/vexflow/wiki
- update READMEs with important information
- generate documentation at https://github.com/vexflow/vexflow-docs/ 
  - let's do it in a versioned directory so we can save the API docs of important releases. (e.g. `vexflow-docs/5.0/`, `vexflow-docs/5.1.2/`). The current version can always be in `vexflow-docs/current/`

Lets avoid behavior changing PRs and focus on bug fixes and general repo polish (e.g., comments) during this beta testing period.

If you have any improvements that result in behavioral changes, please propose in the discussions: https://github.com/vexflow/vexflow/discussions

The SVG issues will be solved after 5.0, for the 5.1 release.

**Comment by rvilarl** [comment]
@ronyeh are you ok with shipping a beta 1?

**Comment by ronyeh** [comment]
Hi, sorry I've been busy. I think I'll have more time to work on VexFlow in early 2025.

I think before shipping v5.0, there should be more documentation / examples, and a simple sandbox so that people can explore the APIs more quickly.

It's OK that there are backwards incompatibilities and (temporary, I hope) regressions. But it is important that we can highlight the new benefits and allow newcomers to learn the API quickly (via wiki docs, easy examples, and the sandbox).

I'd be happy to look into the sandbox / playground for v5.

**Comment by rvilarl** [comment]
Looking forward to seeing us back to coding.

**Comment by rvilarl** [comment]
@ronyeh vexflow5 is now integrated on pianoplay https://michaelecke.com/pianoplaynew/ still not functional but fine from a representation perspective.

**Comment by rvilarl** [comment]
@ronyeh @jaredjj3 I have created a new beta release to continue with the integration of vexflow5, vexml, pianoplay.

**Comment by jaredjj3** [comment]
That's great! [vexml 0.1.4](https://github.com/stringsync/vexml/releases/tag/v0.1.4) now uses vexflow 5.0.0-beta.1. The measure lines look sharper in the newest version, for example:

![Image](https://github.com/user-attachments/assets/a8b8584f-b9b8-4a6a-9d84-d221f3ec916f)

**Comment by rvilarl** [comment]
@ceymard see comment above. Thanks! Your PR is included in vexflow 5.0.0-beta.1

**Comment by ceymard** [comment]
Alright !

**Comment by rvilarl** [comment]
What about releasing 5.0? I am happy with the vexml / pianoplay integration. As @mscuthbert stated july last year "In the meantime, contributions to other parts of the library (formatting, layouts, beam improvements) are languishing. "

I think that we need feedback from users.

@mscuthbert @0xfe @AaronDavidNewman @ronyeh @jaredjj3 what do you think?

**Comment by mscuthbert** [comment]
@rvilarl -- you already have my quote from 7-8 months ago :-)  I still think the same thing only more urgent -- I have a lot of VF patches I want to make regarding beam quality and layout but Artusi's not going to move from 0xfe repo to VF5 on this repo before there's a release.  Thanks!

**Comment by ronyeh** [comment]
@rvilarl I don't want to hold up your launch since you and Jared @jaredjj3 are the biggest consumers of V5. If both of you give the thumbs up and are OK with going forward, let's get this tested and launched. We can get it shipped in March, as long as we aren't merging any new big features. (Only important bug fixes are allowed from here on out.)



The only things I wanted to do were:
1) improve documentation (mostly on the wiki: https://github.com/vexflow/vexflow/wiki)
2) improve example code that is hosted inside the main repo. The more simple/streamlined examples the better, since these examples can be used as context for an AI-assisted coding environment.
3) make a sandbox so people can test their JS or ideally TS code and see the results immediately.

Neither of these are blockers for the release, but they are definitely nice to have.

Post release, we can review and merge all the new PRs so they can go into 5.1. We can also investigate the SVG PR.

**Comment by jaredjj3** [comment]
@ronyeh, sure I'll add my vote!

As you know, I’m not necessarily representative of the typical VexFlow user (evidenced by a limited number of MusicXML rendering libraries for the web). vexml is currently in a pre-v1 state, so my requirements are flexible and I don’t expect my vote to carry significant weight. That said, I do have a personal stake in seeing v5 move forward in a way that strengthens the community.

For me, the requirement I care most about is clear font-loading instructions that are easily discoverable. I see there are already examples on [vexflow-examples](https://vexflow.github.io/vexflow-examples/getting-started/), but it requires some digging. A v4 to v5 migration guide might be a better way to surface this information for users upgrading.

While a major version bump does indeed communicates breaking changes, I think it would be beneficial to provide an opt-out escape hatch for v4 users who want to upgrade to v5 but aren’t ready for all the changes. This could be phased out in v6, reducing friction while still allowing the project to evolve. Does this already exist in the current v5-beta?

Ultimately, I want v5 to reinvigorate the VexFlow community, not split it. Striking a balance between ease-of-use, technical debt, and performance is critical. I appreciate the effort being made to ensure that v5 is a meaningful step forward.

**Comment by ronyeh** [comment]
Thanks Jared!

I agree that clear documentation, more examples, and a V4 vs V5 migration guide (and/or comparison table) are the most important things to have for a nice new major release.

I don't believe font loading is huge issue if you happen to use the default bundle, which is a `vexflow.js` file built for common JS environment (e.g., `<script></script>` tag). The default bundle includes the same three fonts that we have been including: Bravura, Petaluma, Gonville.

The main complication I see is that we have been supporting a flexible dev environment. Different developers have different preferences for how to load vexflow, and in V3, V4, and now V5 we are basically supporting all of them:
- CJS (the V3 way. It's just a vexflow.js file that you can include via script tag.)
- ESM (introduced in V4. Allows modern JS bundlers to load the right modules on the fly, or roll up a smaller bundle if you don't use all the modules.)
- TypeScript (introduced in V4. You build in a TS environment and it outputs your app with webpack/esbuild or another bundler).

Additionally we now have collected different configurations for which fonts we support and load:
- V1 just included Gonville
- The latest version of V3 included Bravura, Petaluma, and Gonville.
- V4 introduced support for more SMuFL fonts, and the idea that you might not want to have a bundle with ALL THE FONTS. So now you can build a vexflow-bravura.js or a vexflow-petaluma.js or a vexflow-leland.js. But it was cumbersome because the developer had to extract all the paths of their desired SMuFL font and adjust all the metrics ahead of time.
- V5 is a major change in that we no longer extract the glyph paths from the font files ahead of time (before even building your app). The glyphs are now displayed at runtime with the fonts, loaded by the browser at runtime. This allows you to use all the SMuFL fonts.

The two gotchas I know of are 1) make sure you include a `<meta charset="utf-8">` in your pages to avoid weird looking glyphs, and 2) if you rely on exported SVG files, you need to stick with V4 for now.

* * * * *

The above sounds complicated, and it is. But the default path should be to have a simple example that bundles the V3 fonts into your app. Then our documentation can point out the different options... maybe through some wiki-choose-your-own-adventure-style decision tree ("Do you use TypeScript?" > "Which font do you want?" > "Do you want the option of more than one font?")

V4 escape hatch? Unfortunately there isn't one for certain situations. If you rely on using SVG paths, V5 doesn't support it because it just sticks a SMuFL glyph directly into a SVG text element. We plan to investigate this *after* launching V5. If we require backwards compatibility, then this would be a huge blocker for V5 getting out the door.


So... stepping back from all the above details, the nice thing is that (I believe) the default basic demos of VexFlow still look the same (V3 > V4 > V5). We haven't changed too many of the standard APIs so example code will mostly look the same and be easy to migrate.

We just need to get around to building a migration table. Maybe something like:
| V5  | V4 | V3 |
| ------------- | ------------- | ------------- |
| V5 snippet 1  | V4 snippet 1  | V3 snippet 1  |
| V5 output 1  | V4 output 1  | V3 output 1  |
| V5 snippet 2  | V4 snippet 2  | V3 snippet 2  |
| V5 output 2  | V4 output 2  | V3 output 2  |
| V5 snippet 3  | V4 snippet 3  | V3 snippet 3  |
| V5 output 3  | V4 output 3  | V3 output 3  |
| V5 snippet 4  | V4 snippet 4  | V3 snippet 4  |
| V5 output 4  | V4 output 4  | V3 output 4  |
| V5 snippet 5  | V4 snippet 5  | V3 snippet 5  |
| V5 output 5  | V4 output 5  | V3 output 5  |


Feel free to ask us any more questions. We definitely need a good `Getting Started` tutorial and a `V4 vs V5` document (the basic example snippets should look the same).

**Comment by ronyeh** [comment]
> As you know, I’m not necessarily representative of the typical VexFlow user...

Well, as far as V5 is concerned, I am treating you (@jaredjj3), @rvilarl, and @AaronDavidNewman as the primary target users/developers. What you say goes. :-)

V4 was used by @mscuthbert in music21j after we migrated from an earlier major version.

And OSMD uses a customized version of VexFlow 1.2.93.

**Comment by rvilarl** [comment]
Thanks Ron. I would take v3 out of the table. Documenting v5 vs v4 hope is  enough.

**Comment by jaredjj3** [comment]
@ronyeh, thanks for the thorough explanation! It sounds like it would require a huge amount of effort to support multiple dev environments. I do think it makes sense to release then.

I support V5 moving forward in hopes of getting more feedback. I don't think we need to block on the incremental improvements.

**Comment by rvilarl** [comment]
I will release in about 3 hours then. Looking forward to getting feedback from users. :)

**Comment by rvilarl** [comment]
Finally done thanks everyone for your support! Let us see what happens now.

**Comment by judocode** [comment]
Hi all, user here, how do I upgrade? 😄 I used to be able to do `import Vex from 'vexflow/gonville'` but gonville seems no longer there, so I just went with the default `import Vex from 'vexflow'` and I used to be able to do things like `export type VexNote = Vex.BarNote | Vex.StaveNote | Vex.TabNote`, but it's now saying `Vex` is a type and not a namespace.

**Comment by rvilarl** [comment]
Please use VexFlow with Vex. (ie.; VexFlow.StaveNote)
